WordPress/Classes/Services/NullBlogPropertySanitizer.swift

```diff
@@ -0,0 +1,66 @@
+import Foundation
+
+/// Delete invalid rows in the database whose required blog properties are NULL
+///
+/// This is born from the investigation of the crash described in #12028 (https://git.io/JePdZ).
+/// Unfortunately, we could not find the cause of it. The findings are described in that issue.
+///
+/// This tries to “fix” the issue by deleting the orphaned entities like `Post` whose `blog` were
+/// set to NULL. They are currently inaccessible because they are not attached to any `Blog`. But
+/// leaving them in the database would cause a crash if Core Data tries to save new entities.
+///
+struct NullBlogPropertySanitizer {
+    private let store: UserDefaults
+    private let key = "null-property-sanitization"
+
+    private let context: NSManagedObjectContext
+
+    init(store: UserDefaults = UserDefaults.standard,
+         context: NSManagedObjectContext = ContextManager.shared.mainContext) {
+
+        self.store = store
+        self.context = context
+    }
+
+    func sanitize() {
+        guard appWasUpdated() else {
+            return
+        }
+
+        let entityNamesWithRequiredBlogProperties = [
+            Post.entityName(),
+            Page.entityName(),
+            Media.entityName(),
+            PostCategory.entityName()
+        ]
+
+        context.perform {
+            entityNamesWithRequiredBlogProperties.forEach { entityName in
+                let request = NSFetchRequest<NSManagedObject>(entityName: entityName)
+                let predicate = NSPredicate(format: "blog == NULL")
+                request.predicate = predicate
+
+                if let results = try? self.context.fetch(request), !results.isEmpty {
```

Comment:
Have you considered a [NSBatchDeleteRequest](https://www.avanderlee.com/swift/nsbatchdeleterequest-core-data/) instead? This seems like it could be relatively slow if there are many entities to fix, and it being in the launch path might end up causing a launch timeout.

Reply:
Very good point. The problem with the `NSBatchDeleteRequest` is that it will not use the validation rules we defined in Data Core thus not removing comments (for example). Although I believe this wouldn't be a problem this would leave the database in (another) inconsistent state.

One thing that I did was to move the place when we update the version that the sanitization happened (46df739af9f0f5e44b4ce2df701ce2cc0c46fc96). In a case where this fails or a timeout occurs, the next time it will not be executed — and the user wouldn't get locked out of the app.

We have two options here, both with pros/cons:

I guess we have a few options, let me know which one looks the better in your opinion:

- Use `NSBatchDeleteRequest` which is way quicker but with the risky of leaving the database in another consistent state;
- Keep the regular `delete` which is slower and might cause timeout with huge databases but will keep it consistent

@koke thoughts?

Reply:
> The problem with the NSBatchDeleteRequest is that it will not use the validation rules we defined in Data Core thus not removing comments (for example)

I'm not sure I understand which validation rules are you referring to, is this about the cascading deletes? So if I delete a post with a batch delete it won't get rid of the associated comments? That's pretty bad 😞

Reply:
I just realized that because this uses `context.perform`, the actually work is enqueued and out of the startup sequence, so this should not cause a launch timeout 😌 Optimizing this becomes less important now, but maybe we can add a `request.includesPropertyValues = false` to avoid loading all the data for objects that we are about to delete.

Reply:
> I'm not sure I understand which validation rules are you referring to, is this about the cascading deletes? So if I delete a post with a batch delete it won't get rid of the associated comments? That's pretty bad 😞

It's exactly that. :(

> Optimizing this becomes less important now, but maybe we can add a request.includesPropertyValues = false to avoid loading all the data for objects that we are about to delete.

Added in 9d3fcbdedfef7bafbcf0ab470cf94496c4fefd84